scripts/analyze_training_metrics.py

```diff
@@ -0,0 +1,168 @@
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import argparse
+import json
+import logging
+import sys
+from collections import defaultdict
+from datetime import datetime, timezone
+from pathlib import Path
+from typing import Dict
+
+import matplotlib.dates as mdates
+import matplotlib.pyplot as plt
+
+LOGGER = logging.getLogger(__name__)
+
+logging.basicConfig(level=logging.INFO)
+
+
+def plot_graph(
+    model_name: str,
+    metric_name: str,
+    values_dict: Dict[datetime, float],
+    output_directory: Path,
+) -> bool:
+    sorted_metrics = sorted(values_dict.items())
+    x, y = zip(*sorted_metrics)
+
+    # Compute the threshold
+    if len(y) >= 2:
+        before_last_value = y[-2]
+    else:
+        before_last_value = y[-1]
+    ninety_five_percent = before_last_value * 0.95
+
+    figure = plt.figure()
+    axes = plt.axes()
+
+    # Formatting of the figure
+    figure.autofmt_xdate()
+    axes.fmt_xdata = mdates.DateFormatter("%Y-%m-%d-%H-%M")
+    axes.set_title(f"{model_name} {metric_name}")
+
+    # Display threshold
+    axes.axhline(y=ninety_five_percent, linestyle="--", color="red")
+    plt.annotate(
+        "{:.4f}".format(ninety_five_percent),
+        (x[-1], ninety_five_percent),
+        textcoords="offset points",  # how to position the text
+        xytext=(-10, 10),  # distance from text to points (x,y)
+        ha="center",
+        color="red",
+    )
+
+    # Display point values
+    for single_x, single_y in zip(x, y):
+        label = "{:.4f}".format(single_y)
+
+        plt.annotate(
+            label,  # this is the text
+            (single_x, single_y),  # this is the point to label
+            textcoords="offset points",  # how to position the text
+            xytext=(0, 10),  # distance from text to points (x,y)
+            ha="center",
+        )  # horizontal alignment can be left, right or center
+
+    axes.plot_date(x, y, marker=".", fmt="-")
+    # plt.show(block=True)
+    output_file_path = output_directory.resolve() / f"{model_name}_{metric_name}.svg"
+    LOGGER.info("Saving %s figure", output_file_path)
+    plt.savefig(output_file_path)
+
+    plt.close(figure)
+
+    # Check if the threshold has been crossed
+    return y[-1] < ninety_five_percent
+
+
+def analyze_metrics(metrics_directory: str, output_directory: str):
+    root = Path(metrics_directory)
+
+    metrics = defaultdict(
+        lambda: defaultdict(dict)
+    )  # type: Dict[str, Dict[str, Dict[datetime, float]]]
```

Comment:
Can't the type annotation be actually used here? Or is the defaultdict preventing that?